### PR TITLE
PyTensor intro walkthrough

### LINKED ISSUE #12: Conceptual intro
## What is PyTensor

A library to define and manipulate array-based computational graphs

* Represent numpy-like computations in a graph composed of Python objects
* Manipulate graph at the Python level for:
    * Optimization and stabilization
    * Automatic differentiation
    * Vectorization and "Scalarization" 
    * Anything else: Shape inference, logp inference
* Compile to performant non-python backends: C, Numba, JAX, PyTorch (or just vanilla numpy)

## How does it compare with other frameworks
* Analogy with Sympy
* Graph is specified explicitly and with placeholder inputs
* It is focused on array (tensor) operations (dense and sparse). Narrow/hidden support for other types like scalars, lists, slices, random Generators, strings, None (although easy to extend)
* Side-effects free from the user perspective
* EVERYTHING IS LAZY!
* Trivial support for "partialling" / subgraph extraction.
* Strong focus on hackability / user-defined rewrites
* It's completely ours! (And nobody else uses it)
* Evolved from:
    1. Theano which strongly inspired Tensorflow 1.x and JAX. Many concepts stood the test of time. Others have aged and provide some drag.
  2. Aesara, which cleaned up the codebase, added alternative backends (Numba and JAX) and proved there's some interest out there in a library like this.

### LINKED ISSUE #13: One minute example
Maybe the same we then re-implement in #1 

And/or logistic regression without shared variables?
https://pytensor.readthedocs.io/en/latest/tutorial/examples.html#a-real-example-logistic-regression

### PR DESCRIPTION
**Pull request by ricardoV94**
Closes #12
Closes #13

### CODE REVIEW

**Comment by ricardoV94** [comment]
One thing that may be useful is to wrap up with an applied simple example. Logistic regression with a naive gradient descent?

**Comment by jessegrabowski** [comment]
@ricardoV94 ping for review on your own PR

File: notebooks/walkthrough/pytensor_intro.ipynb [hunk 1]
```diff
@@ -0,0 +1,2873 @@
+{
```
Comment (ricardoV94) [inline notebooks/walkthrough/pytensor_intro.ipynb:1]: <p>Typo: hackabel</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pymc-devs/pytensor-workshop/pull/34/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='2'/>

Comment (ricardoV94) [inline notebooks/walkthrough/pytensor_intro.ipynb:1]: <p>fused log add is incorrect (as we'll eventually mention fusion optimizations one way or another). The owner is log, which takes as input the output of an add operation</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pymc-devs/pytensor-workshop/pull/34/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='15'/>

Comment (ricardoV94) [inline notebooks/walkthrough/pytensor_intro.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #2.    y = np.exp(x)
</pre><p>Do you want to use <code>np.exp</code>?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pymc-devs/pytensor-workshop/pull/34/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='31'/>

Comment (ricardoV94) [inline notebooks/walkthrough/pytensor_intro.ipynb:1]: <p>I wouldn't use the word <code>side effects</code>.Perhaps say it will change the meaning of the graph_  </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pymc-devs/pytensor-workshop/pull/34/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='33'/>

Comment (ricardoV94) [inline notebooks/walkthrough/pytensor_intro.ipynb:1]: <p>Should print shape to show something changed, and I would call <code>x</code>,<code>x_vec</code>     </p><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pymc-devs/pytensor-workshop/pull/34/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='38'/>

Comment (ricardoV94) [inline notebooks/walkthrough/pytensor_intro.ipynb:1]: <p>A bit odd to have this between the vectorization and scalarization?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pymc-devs/pytensor-workshop/pull/34/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='41'/>

Comment (ricardoV94) [inline notebooks/walkthrough/pytensor_intro.ipynb:1]: <p>Vectorization was one such case of graph-to-graph-transformation as well?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pymc-devs/pytensor-workshop/pull/34/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='48'/>

Comment (ricardoV94) [inline notebooks/walkthrough/pytensor_intro.ipynb:1]: <p>loop lifting -&gt; loop fusion</p><p> </p><p>More interesting is <code>cos</code> writing directly on the squeezed output. In numpy this would be doing the optimiztaino <code>np.cos(x, out=x)</code> which quickly becomes unreadable. Also squeeze is a view of CGemv so no intermediate allocation there either, but we can omit that.  If you want to show use <code>print_memory_map</code> in the new version that shows both the destroy and view maps </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pymc-devs/pytensor-workshop/pull/34/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='72'/>

Comment (ricardoV94) [inline notebooks/walkthrough/pytensor_intro.ipynb:1]: <p>typo audiance</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pymc-devs/pytensor-workshop/pull/34/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='78'/>

Comment (ricardoV94) [inline notebooks/walkthrough/pytensor_intro.ipynb:1]: <p>The use of two functions inadvertently highlights one of the advantages of the pytensor graph-way of building computations. You don't need to think in advance how to define intermediate functions so that you can get intermediate outputs from specific inputs to JIT (as you would need to in Numba/JAX). You just grab the relevant variable and plug them as inputs/outputs. May be worth to mention. </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pymc-devs/pytensor-workshop/pull/34/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='87'/>

Comment (ricardoV94) [inline notebooks/walkthrough/pytensor_intro.ipynb:1]: <p>There's no second step, we just pass <code>exp(new_input)</code> instead of <code>old_input</code> . I just called that <code>log_x</code> because once you exponentiate it you get <code>x</code> (old input) </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pymc-devs/pytensor-workshop/pull/34/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='58'/>

**Comment by ricardoV94** [comment]
Looks good @jessegrabowski , left some small notes

File: notebooks/walkthrough/pytensor_intro.ipynb (hunk 1 above)
Comment (jessegrabowski) [reply notebooks/walkthrough/pytensor_intro.ipynb:1]: <p>Nope!</p>

Comment (jessegrabowski) [reply notebooks/walkthrough/pytensor_intro.ipynb:1]: <p>I think it's too big brain, I simplified it.</p><p> </p>

File: notebooks/walkthrough/pytensor_intro.ipynb [hunk 2]
```diff
@@ -0,0 +1,3156 @@
...
```
Comment (ricardoV94) [inline notebooks/walkthrough/pytensor_intro.ipynb:1]: <p>This cell should be moved under the next header</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pymc-devs/pytensor-workshop/pull/34/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='44'/>
